src/chainstate/stacks/miner.rs

```diff
@@ -469,6 +474,9 @@ impl<'a> StacksMicroblockBuilder<'a> {
         let mut num_selected = 0;
         let deadline = get_epoch_time_ms() + (self.settings.max_miner_time_ms as u128);
 
+        mem_pool.reset_last_known_nonces()?;
```

Comment:
Won't this obliterate the known nonces for the *entire* mempool?  Isn't that going to make microblock mining expensive, if we have to re-calculate nonces all the time?  Or, is that bound by the subsequent call to `estimate_tx_rates` with `max_updates = 100`?

Reply:
Yes, it will obliterate the nonces for the entire mempool. However, it doesn't make mining any more expensive (in terms of nonce lookups) than the current implementation -- nonce calculation is done lazily. When a transaction is considered for inclusion, if the expected nonces aren't known, the origin and sponsor nonces are looked up, and the database is updated with those nonces, then the transaction is thrown back in for consideration, and the iterator polls for the next transaction to be considered (which could be the one it just saw, if it was valid given the expected nonce). That is:

1. Get the transaction from the mempool with the highest fee rate whose expected nonce is not known _or_ the transaction nonce matches the known expected nonce.
2. If the transaction's nonce was not known, look it up and update the table, go back to 1.
3. Otherwise, try to add the transaction to the block, then bump the expected nonce, go back to 1.